athena-google-bigquery/src/test/java/com/amazonaws/athena/connectors/google/bigquery/BigQueryRecordHandlerTest.java

```diff
@@ -256,6 +276,56 @@ public void getObjectFromFieldValue()
         }
     }
 
+    @Test
+    public void testReadComplexTypes()
+            throws Exception
+    {
+        try (ReadRecordsRequest request = new ReadRecordsRequest(
+                federatedIdentity,
+                BigQueryTestUtils.PROJECT_1_NAME,
+                "queryId",
+                new TableName("dataset1", "table1"),
+                BigQueryTestUtils.getComplexTypeTestSchema(),
+                Split.newBuilder(S3SpillLocation.newBuilder()
+                                .withBucket(bucket)
+                                .withPrefix(prefix)
+                                .withSplitId(UUID.randomUUID().toString())
+                                .withQueryId(UUID.randomUUID().toString())
+                                .withIsDirectory(true)
+                                .build(),
+                        keyFactory.create()).build(),
+                new Constraints(Collections.EMPTY_MAP),
+                0,          //This is ignored when directly calling readWithConstraints.
+                0)) {   //This is ignored when directly calling readWithConstraints.
+            //Always return try for the evaluator to keep all rows.
+
+            //Populate the schema and data that the mocked Google BigQuery client will return.
+            com.google.cloud.bigquery.Schema tableSchema = BigQueryTestUtils.getComplextTypeTestSchema();
+            List<FieldValueList> tableRows = Arrays.asList(
+                    BigQueryTestUtils.getBigQueryComplexTypeFieldValueList(ImmutableList.of(1, 2, 3), 1),
+                    BigQueryTestUtils.getBigQueryComplexTypeFieldValueList(ImmutableList.of(10, 20, 30), 2),
+                    BigQueryTestUtils.getBigQueryComplexTypeFieldValueList(ImmutableList.of(100, 200, 300), 3),
+                    BigQueryTestUtils.getBigQueryComplexTypeFieldValueList(ImmutableList.of(1000, 2000, 3000), 4),
+                    BigQueryTestUtils.getBigQueryComplexTypeFieldValueList(ImmutableList.of(10000, 20000, 30000), 5),
+                    BigQueryTestUtils.getBigQueryComplexTypeFieldValueList(ImmutableList.of(100000, 200000, 300000), 6)
+            );
+            Page<FieldValueList> fieldValueList = new BigQueryPage<>(tableRows);
+            TableResult result = new TableResult(tableSchema, tableRows.size(), fieldValueList);
+
+            //Mock out the Google BigQuery Job.
+            Job mockBigQueryJob = mock(Job.class);
+            when(mockBigQueryJob.isDone()).thenReturn(false).thenReturn(true);
+            when(mockBigQueryJob.getQueryResults()).thenReturn(result);
+            when(bigQuery.create(nullable(JobInfo.class))).thenReturn(mockBigQueryJob);
+
+            QueryStatusChecker queryStatusChecker = mock(QueryStatusChecker.class);
+            when(queryStatusChecker.isQueryRunning()).thenReturn(true);
+            //Execute the test
+            bigQueryRecordHandler.readWithConstraint(spillWriter, request, queryStatusChecker);
```

Comment:
Great job on adding test and ensuring no errors or exceptions are being thrown during function call! However, it'd be great if the output of `readWithConstraint` also be tested.

Reply:
@akuzin1, the only output from this void method (readWithConstraint) we can test is to check to see if the records are being spilled. This assertion is duly added. Thank you for pointing this out.